### DISCUSSION TITLE
Real time simulation crashes after a few hours

### DISCUSSION THREAD

**Discussion by MidrarAdham** [GRIDAPPSD/gridappsd-forum#151] [Q&A]
Hi everyone,

I use the IEEE 9500 model to simulate DER operations in real time. The simulation crashes after some time (depending on the GridAPPS-D version).

## Model Set up:

- I removed all existing DERs from the 9500 model and inserted 2549 residential-scale DERs instead.
- No other changes were made to the model.

## DERs control values:

- The DERs power consumption profiles are sent to the DER models within the feeder.
- The profiles I am using are in one-minute timestep resolution.

## The Issue:

- The model works fine, at least for the first two hours.
- Then, the simulation crashes with multiple errors. Please see the attached simulation log files.
[gridappsd.log](https://github.com/user-attachments/files/17963174/gridappsd.log)
[gridappsd_1.log](https://github.com/user-attachments/files/17963175/gridappsd_1.log)
[gridappsd_simulation_logs_from_terminal.txt](https://github.com/user-attachments/files/17963176/gridappsd_simulation_logs_from_terminal.txt)

## PC Setup:
- I am using GridAPPS-D on WSL Ubuntu
- RAM 32GB, 6-core AMD Processor
- Docker allocated memory is 16GB

## Attempted Solutions:

- I tried changing the GridAPPS-D versions by checking out different GridAPPS-D releases (latest: 2024.09.0, 2022.06, 2021.09), but the problem persists.
- In the 2024 version, the simulation crashes after 15 minutes. The older versions run for multiple hours.
- At first, I thought upgrading the PC might solve the issue. However, I monitored the memory and CPU behavior during a simulation and found that the memory usage increased as the simulation went through, suggesting a potential memory leakage within the gridappsd container (please see the following figure)
- The fallout at the end of the plot is when the simulation crashes.

![docker_memory](https://github.com/user-attachments/assets/00847c9a-e7ec-4d17-9581-2ce37b4474f9)

## Questions:

- Do you happen to know if this issue has been encountered before? If so, is there another version where the issue was addressed?
- If not, are there any suggestions as to what to do or where to start in order to debug this?
Maybe I am viewing this all wrong. Please let me know if there is something I missed.

Thank you!

**Comment by poorva1209** [comment]
@afisher1 Can you have a look at this? I saw the logs and it shows following error after running for more than 2 hours. We can convert this into an issue on board if it requires any update to the platform. 

```
2024-11-29 21:11:37,961 | INFO  | Thread-39        | LogManagerImpl                   | 57 - gov.pnnl.goss.gridappsd.gridappsd - 0.0.1.202201120004 | 1732914697961|helics_goss_bridge.py|2096178979|RUNNING|system|INFO
incrementing to 9944

2024-11-29 21:11:37,970 | WARN  | Thread-39        | LogManagerImpl                   | 57 - gov.pnnl.goss.gridappsd.gridappsd - 0.0.1.202201120004 | 1732914697970|helics_goss_bridge.py|2096178979|RUNNING|system|WARN
Simulation 2096178979 is running slower than real time!!!. Time step took 20.335416546004126 seconds to execute

2024-11-29 21:11:37,973 | ERROR | Thread-39        | LogManagerImpl                   | 57 - gov.pnnl.goss.gridappsd.gridappsd - 0.0.1.202201120004 | 1732914697973|helics_goss_bridge.py|2096178979|ERROR|system|ERROR
The HELICS co-simulation for simulation 2096178979 entered an error state for some unknown reason.

2024-11-29 21:11:37,980 | ERROR | Thread-39        | LogManagerImpl                   | 57 - gov.pnnl.goss.gridappsd.gridappsd - 0.0.1.202201120004 | 1732914697980|helics_goss_bridge.py|2096178979|ERROR|system|ERROR
Error in run simulation Traceback (most recent call last):
  File "/gridappsd/services/helicsgossbridge/service/helics_goss_bridge.py", line 532, in run_simulation
    raise RuntimeError(f"The HELICS co-simulation for simulation {self._simulation_id} entered an error state for some unknown reason.")
RuntimeError: The HELICS co-simulation for simulation 2096178979 entered an error state for some unknown reason.```

**Reply by afisher1** [reply]
@poorva1209 I think this is related to the memory issue Craig is trying to track down.

**Reply by poorva1209** [reply]
Could be...I am looking into it.

**Reply by MidrarAdham** [reply]
Hi everyone,

I've been working on this issue and I suspect that the memory leak is coming from the GridLAB-D source code. However, in order to confirm this, I need to use the adjusted GridLAB-D source code and link it to GridAPPS-D. 


- Are there instructions as to how GridLAB-D and GridAPPS-D linked? I looked at the [documentation](https://gridappsd-training.readthedocs.io/en/develop/installation/1.8-Platform-Release-History.html) and I could not find information there.
- I made some adjustments that address some potential memory leak issues within GridLAB-D source code. Is it possible to test the adjusted source code within GridAPPS-D? If so, how?

Thank you all again for your input.

**Reply by poorva1209** [reply]
Sure, I will add the instructions here shortly for building GridLAB-D within the container.

**Reply by poorva1209** [reply]
Below are the steps we are using to build GridLAB-D within the container. The file that builds the base GridAPPS-D container is here: https://github.com/GRIDAPPSD/gridappsd-docker-build/blob/1155323a8b66ca3f58145d082561b87c5a90bbfc/Dockerfile.gridappsd_base#L140. You can just make your changes within the container that you are running. 

You can replace the git clone step with your own source code within the container. **Please note that we are using "gridappsd-branch" of the gridlab-d repository. It would be best that you make your changes locally using that branch to make sure we are using the same version.** 

```
RUN cd $TEMP_DIR \
    && git clone https://github.com/gridlab-d/gridlab-d.git -b gridappsd-branch --single-branch \
    && cd ${TEMP_DIR}/gridlab-d/third_party \
    && tar -xzf xerces-c-3.2.0.tar.gz \
    && cd ${TEMP_DIR}/gridlab-d/third_party/xerces-c-3.2.0 \
    && ./configure \
    && make \
    && make install \
    && cd ${TEMP_DIR}/gridlab-d \
    && autoreconf -if \
    && ./configure --with-helics=/usr/local/bin --prefix=$GLD_INSTALL --with-fncs=$FNCS_INSTALL --enable-silent-rules 'CFLAGS=-g -O2 -w' 'CXXFLAGS=-g -O2 -w -std=c++14' 'LDFLAGS=-g -O2 -w' \
    && make \
    && make install \
    && cd /tmp \
    && /bin/rm -r ${TEMP_DIR}/gridlab-d
```

**Reply by MidrarAdham** [reply]
Great, that's good to know about the gridappsd-branch. I'll look into it. Thank you!

**Reply by MidrarAdham** [reply]
Hi Poorva,

As I was trying to set up a local instance of GridAPPS-D, the GOSS-GridAPPS-D repository returned an error, because it does not include a "dockerized" branch. Looking at the Dockerfile.gridappsd in the develop and some other branches, it tries to clone the GOSSS-GridAPPS-D repository with a branch called "dockerized" ([see here](https://github.com/GRIDAPPSD/gridappsd-docker-build/blob/1155323a8b66ca3f58145d082561b87c5a90bbfc/Dockerfile.gridappsd))

I tried different branches (develop and releases/2023.07.0), however, it seems that the ones I tried use the dockerized branch as well. I guess my question is, where can I find the "dockerized" branch of the GOSS-GridAPPS-D repository? 

Thank you.

**Comment by shpoudel** [comment]
Hi Midrar,

Good to connect with you.

While our software/platform team is investigating this, I would like to know more details on the modeling side of the modified test feeder.

What are the attributes of those 2000+ DERs added to the system, and what does the power consumption profile look like? The system may be constrained by the 2000+ DER dispatches.

**Reply by MidrarAdham** [reply]
Hi Shiva,

Always! The profiles use real and reactive power values. However, the reactive power values are **NOT** sent to the DERs; only the real power values are sent to the platform. The maximum real power value across all profiles is <= 4.7 kW. I attached one of the profiles to this message. I will reduce the number of DERs incrementally and see if that changes anything.
[ders_0.csv](https://github.com/user-attachments/files/17983657/ders_0.csv)

Thank you, everyone, for the prompt responses.

**Reply by shpoudel** [reply]
And, I guess you have allocated those DERs to different service transformers in the circuit, right?

**Reply by MidrarAdham** [reply]
Correct.
